frontend/model/notifications/periodicNotifications.js

```diff
@@ -38,59 +34,142 @@ const validateNotificationData = objectOf({
+ *        - Marks the notification as fired in the `firedMap`.
+ *    - If the notification has been fired and its clear condition evaluates to
+ *      true, it:
+ *        - Removes the notification from the `firedMap`.
+ * 3. Updates the `lastRun` timestamp to the current time.
+ * 4. Sets a timeout to recursively call `runNotificationListRecursive` after
+ * the specified delay, adjusting for the time that has already passed since the
+ * last run.
+ */
+async function runNotificationListRecursive () {
   const rootState = sbp('state/vuex/state')
   const rootGetters = sbp('state/vuex/getters')
-  const firedMap = rootState.periodicNotificationAlreadyFiredMap
-  const callWithStates = func => func.call(data.state, { rootState, rootGetters })
+  const firedMap = rootState.periodicNotificationAlreadyFiredMap.alreadyFired
+  const lastRunMap = rootState.periodicNotificationAlreadyFiredMap.lastRun
+  const callWithStates = (func, stateKey) => func.call(ephemeralNotificationState.partition[stateKey], { rootState, rootGetters })
 
-  for (const entry of data.notifications) {
-    if (!firedMap[entry.stateKey] && callWithStates(entry.emitCondition)) {
-      await callWithStates(entry.emit)
-      Vue.set(firedMap, entry.stateKey, true)
-    }
+  // Exit if a timeout is already set
+  if (ephemeralNotificationState.clearTimeout) return
+  // Prevent accidentally calling `runNotificationListRecursive` while it's
+  // still running
+  // We set `clearTimeout` to a local value to stop processing notifications
+  // when clearing stop notifications, to avoid calling
+  // `runNotificationListRecursive` while it's running and to be able to compare
+  // it to the local value later to replace `clearTimeout` later on in this
+  // function
+  let aborted: boolean = false
+  const noop = () => { aborted = true }
+  ephemeralNotificationState.clearTimeout = noop
+
+  // If there are any queued invocations, wait until they're done
+  await Promise.all(
+    ((Object.entries(sbp('okTurtles.eventQueue/queuedInvocations')): any): [string, (Function | string[])[]])
+      .map(([queue, invocations]) => {
+        return !!invocations.length && sbp('okTurtles.eventQueue/queueEvent', queue, () => {}).catch(() => {})
+      })
+  )
+
+  // Check if enough time has passed since the last run
+  for (const entry of ephemeralNotificationState.notifications) {
+    if (aborted) break
+    const stateKey = entry.stateKey
+    const lastRun = lastRunMap[stateKey] || 0
+    // Use the delay from the notification entry
+    const delay = entry.delay
+    if ((Date.now() - lastRun) >= delay) {
+      try {
+        if (!firedMap[stateKey] && callWithStates(entry.emitCondition, stateKey)) {
+          await callWithStates(entry.emit, stateKey)
+          firedMap[stateKey] = true
+        }
 
-    if (firedMap[entry.stateKey] && callWithStates(entry.shouldClearStateKey)) {
-      Vue.delete(firedMap, entry.stateKey)
```

Comment:
I see multiple places where `Vue.delete` and `Vue.set` are replaced with pure Javascript logics here. I think it is fine because `alreadyFired` and `lastRun` maps are currently only used in this file and _[1] there is no place in the app to re-render in response to these state updates_, as far as I know. For the same reason stated in [1], I think the entire `periodicNotificationAlreadyFiredMap` object can be moved out of Vuex store too but it's not necessary.

Reply:
I agree with you that it's not used elsewhere in the UI AFAIK. The reason to remove `Vue.set` and friends is so that it can run in the SW without having to pull in Vue there as a dependency. However, I think it's fine to leave it in the state (as is) so that we can restore the state alongside the rest of it.